File: src/renderer/camera.cpp
//
// Created by andyroiiid on 6/15/2021.
//

#include "renderer/camera.h"

#include <glm/ext/matrix_clip_space.hpp>

Camera &Camera::getInstance() {
    static Camera instance;
    return instance;
}

void Camera::framebufferResize(int width, int height) {
    screenWidth = static_cast<float>(width);
    screenHeight = static_cast<float>(height);
    matrixDirty = true;
}

void Camera::update() {
    if (matrixDirty) {
        halfWidth = screenWidth / screenHeight * halfHeight;
        shaderGlobals.setMatrix(glm::ortho(
                center.x - halfWidth,
                center.x + halfWidth,
                center.y - halfHeight,
                center.y + halfHeight
        ));
        matrixDirty = false;
    }
}

std::tuple<float, float> Camera::screenToCanvas(int x, int y) const {
    return std::tuple<float, float>(
            static_cast<float>(x),
            screenHeight - static_cast<float>(y)
    );
}

std::tuple<float, float> Camera::screenToWorld(int x, int y) const {
    float normalizedX = static_cast<float>(x) / screenWidth * 2.0f - 1.0f;
    float normalizedY = 1.0f - static_cast<float>(y) / screenHeight * 2.0f;
    return std::make_tuple(
            center.x + normalizedX * halfWidth,
            center.y + normalizedY * halfHeight
    );
}

void Camera::setHalfHeight(float newHalfHeight) {
    halfHeight = newHalfHeight;
    matrixDirty = true;
}

void Camera::setCenter(float x, float y) {
    center = {x, y};
    matrixDirty = true;
}

File: samples/multi_games/camera.lua
local Object = require("classic.lua")

local Camera = Object:extend()

function Camera:new()
    self.halfHeight = 100
    self.halfWidth = 100
    self.x = 0
    self.y = 0
end

function Camera:update()
    local width, height = haru.window.getFramebufferSize()
    self.halfWidth = width / height * self.halfHeight
end

function Camera:preDraw()
    haru.renderer.setMatrixOrtho(
        self.x - self.halfWidth,
        self.x + self.halfWidth,
        self.y - self.halfHeight,
        self.y + self.halfHeight
    )
end

function Camera:setHalfHeight(halfHeight)
    self.halfHeight = halfHeight
end

function Camera:setCenter(x, y)
    self.x = x
    self.y = y
end

function Camera:getMousePosition()
    local x, y = haru.mouse.normalizedPosition()
    x = x * 2 - 1
    y = y * 2 - 1
    return self.x + x * self.halfWidth, self.y + y * self.halfHeight
end

camera = Camera()
